examples/rl/DDPG_Pendulum.py

```diff
@@ -0,0 +1,414 @@
+accumulated so far.
+
+Now, lets see how is it implemented.
+"""
+
+# We use openai gym for Pendulum Env.
+import gym
+import tensorflow as tf
+from tensorflow.keras import layers
+import numpy as np
+import matplotlib.pyplot as plt
+
+"""
+We use [OpenAIGym](http://gym.openai.com/docs) to create the environment.
+We will use the `upper_bound` parameter to scale our actions later.
+"""
+
+problem = "Pendulum-v0"
+env = gym.make(problem)
+
+num_states = env.observation_space.shape[0]
+print("Size of State Space ->  {}".format(num_states))
+num_actions = env.action_space.shape[0]
+print("Size of Action Space ->  {}".format(num_actions))
+
+upper_bound = env.action_space.high[0]
+lower_bound = env.action_space.low[0]
+
+print("Max Value of Action ->  {}".format(upper_bound))
+print("Min Value of Action ->  {}".format(lower_bound))
+
+"""
+To implement Actor exploration, we use noisy perturbations, specifically
+an **Ornstein-Uhlenbeck process**, as described in paper.
+It samples noise from a correlated normal distribution.
+"""
+
+
+class OUActionNoise:
+    def __init__(self, mean, std_deviation, theta=0.15, dt=1e-2, x0=None):
+        self.theta = theta
+        self.mean = mean
+        self.std_dev = std_deviation
+        self.dt = dt
+        self.x0 = x0
+        self.reset()
+
+    def __call__(self):
+        # Formula taken from https://www.wikipedia.org/wiki/Ornstein-Uhlenbeck_process.
+        x = (
+            self.x_prev
+            + self.theta * (self.mean - self.x_prev) * self.dt
+            + self.std_dev * np.sqrt(self.dt) * np.random.normal(size=self.mean.shape)
+        )
+        # This makes this noise more correlated
+        self.x_prev = x
+        return x
+
+    def reset(self):
+        if self.x0 is not None:
+            self.x_prev = self.x0
+        else:
+            self.x_prev = np.zeros_like(self.mean)
+
+
+"""
+The `Buffer` class implements Experience Replay.
+
+---
+![Algorithm](https://i.imgur.com/mS6iGyJ.jpg)
+---
+
+
+**Critic loss** - Mean Squared Error of `y - Q(s, a)`
+where `y` is the expected action value as seen by the Target network,
+and `Q(s, a)` is action value predicted by the Critic network. `y` is a moving target that the critic model tries to achieve; we make this target
+stable by updating the target model slowly.
+
+**Actor loss** - This is computed using the mean of the value given by the Critic network
+for the actions taken by the Actor network. We seek to maximize this quantity.
+
+Hence we update the Actor Network such that it produces actions that get
+the maximum predicted value as seen by the Critic, for a given state.
+
+"""
+
+
+class Buffer:
+    def __init__(self, buffer_capacity=100000, batch_size=64):
+
+        # Number of "experiences" to store at max
+        self.buffer_capacity = buffer_capacity
+        # Num of tuples to train on.
+        self.batch_size = batch_size
+
+        # Its tells us num of times record() was called.
+        self.buffer_ctr = 0
```

Comment:
Again, prefer using fully-spelled out variable names such as "counter", to improve readability